internal/server/server.go

```diff
@@ -108,19 +198,92 @@ func (s *Server) handleLogin() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		session := s.getSession(r)
 
-		rs, err := random.SecureString(32)
+		randomStr, err := random.SecureString(32)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for state")
+			return
+		}
+
+		// generate nonce
+		nonce, err := random.SecureString(32)
 		if err != nil {
-			logAndError(w, http.StatusInternalServerError, err, "error generating random string")
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for nonce")
 			return
 		}
 
-		session.Values["rand"] = rs
+		session.Values["nonce"] = nonce
 
 		if err := session.Save(r, w); err != nil {
 			logAndError(w, http.StatusInternalServerError, err, "error saving session")
 			return
 		}
 
-		fmt.Fprintf(w, "session saved")
+		http.Redirect(w, r, s.oauth2Config.AuthCodeURL(randomStr, oidc.Nonce(nonce)), http.StatusSeeOther)
+	}
+}
+
+func (s *Server) handleCallback() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// handle requests that contain an error
+		if err := r.URL.Query().Get("error"); err != "" {
+			logAndError(w, http.StatusBadRequest, fmt.Errorf(err+": "+r.URL.Query().Get("error_description")), "error in request")
+			return
+		}
+
+		session := s.getSession(r)
+
+		// convert authorization code into an OAuth2 token
+		oauth2Token, err := s.externalFuncs.oauth2ConfigExchange(s.oauth2Config, s.context, r.URL.Query().Get("code"))
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error converting code to token")
+			return
+		}
+
+		// extract the ID token from OAuth2 token
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			logAndError(w, http.StatusInternalServerError, fmt.Errorf("error extracting id_token"), "no id_token in token response")
+			return
+		}
+
+		// verify the ID token
+		idToken, err := s.externalFuncs.oidcIDTokenVerifierVerify(s.verifier, s.context, rawIDToken)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "failed to verify id_token")
+			return
+		}
+
+		// verify the OIDC nonce matches
+		if idToken.Nonce != session.Values["nonce"] {
+			logAndError(w, http.StatusUnauthorized, fmt.Errorf("error invalid nonce"), "id token nonce does not match session nonce")
+			return
+		}
+
+		// extract custom claims
+		var claims struct {
+			Email string `json:"email"`
+		}
+		if err := s.externalFuncs.oidcIDTokenClaims(idToken, &claims); err != nil {
```

Comment:
This shouldnt need a mock function.  Once the `Verifer` returns a proper idtoken with claims, this can use the actual function.

Reply:
This also seems to have the wrong number of args, it only takes a claims interface:

https://pkg.go.dev/github.com/coreos/go-oidc/v3@v3.1.0/oidc#IDToken.Claims

Reply:
Fixed